bootstore/src/db/mod.rs

```diff
@@ -128,16 +173,170 @@ impl Db {
             Ok(())
         })
     }
+
+    pub fn initialize(
+        &self,
+        conn: &mut SqliteConnection,
+        rack_uuid: &Uuid,
+        share_distribution: SerializableShareDistribution,
+    ) -> Result<(), Error> {
+        conn.immediate_transaction(|tx| {
+            if self.is_initialized(tx)? {
+                // If the rack is initialized, a rack uuid must exist
+                let uuid = self.get_rack_uuid(tx)?.unwrap();
+                return Err(Error::AlreadyInitialized(uuid));
+            }
+            let new_uuid = rack_uuid.to_string();
+            match self.initialize_rack_uuid(tx, &new_uuid)? {
+                Some(old_uuid) => {
+                        info!(self.log, "Re-Initializing Rack: Old UUID: {old_uuid}, New UUID: {new_uuid}");
+                    self.update_prepare_for_epoch_0(tx, share_distribution)
+
+                },
+                None => {
+                    info!(self.log, "Initializing Rack for the first time with UUID: {new_uuid}");
+                    self.insert_prepare(tx, 0, share_distribution)
+                    }
+            }
+
+        })
+    }
+
+    // Insert an uncommitted key share for a given epoch This should be called
+    // inside a transaction by the caller so that the given checks can be
+    // performed.
+    fn insert_prepare(
+        &self,
+        tx: &mut SqliteConnection,
+        epoch: i32,
+        share_distribution: SerializableShareDistribution,
+    ) -> Result<(), Error> {
+        use schema::key_shares::dsl;
+        let prepare = KeyShare::new(epoch, share_distribution)?;
+        diesel::insert_into(dsl::key_shares).values(&prepare).execute(tx)?;
+        Ok(())
+    }
+
+    // During rack initialization we set the rack UUID.
+    //
+    // We only allow rewriting this UUID in when the rack is not initialized,
+    // and so we only call this from the `initialize` method.
+    //
+    // Since we only allow a single row, we just delete the existing rows
+    // and insert the new one.
+    //
+    // Return the old rack UUID if one exists
+    //
+    // TODO: Add a DB trigger to ensure only 1 row can exist at a time?
```

Comment:
I'd vote yes, or add a query in this method that confirms after inserting that there is only one. (Assuming that the presence of more than one is Very Bad :tm:)

Reply:
Will do!

Reply:
Done.